Add Microsoft Azure deployment guide (#14733)
metabase/metabase: docs/operations-guide/running-metabase-on-azure.md

__NOTE__: remember that you need to select the region for your Metabase instance according to the location of your users, data warehouse and even the cost of the infrastructure of the region or privacy laws that might restrict cross-border data transfers.

### Step 2: Create the network

Once your resource group is created, click on the name to open the page where you will start creating all your resource. Click on the **Create resources** button at the center of the page and you will be taken to the marketplace menu where you will have to search for the keyword **VNET**:

![VNET Marketplace](images/AZMarketPlaceVnet.png)

From all the offerings, select the VNET product from Microsoft and click on **Create**:

![VNET Product](images/AZVnet.png)

You will be presented with a screen similar to the one you were shown when created the resource group, where you will have to enter a name for the virtual network and select a region for it (remember to use the same region for all the components you create, you don't want your data to be travelling worldwide!) and click on **Next: IP Addresses** button at the bottom.

> Here we will create a simple but secure network topology, composed of a public subnet (a network that is exposed to the internet), and a private network that will contain your application database which will be inaccesible from the outside.

Reviewer: ```suggestion
Here we'll create a secure network composed of a public subnet (a network that is exposed to the internet), and a private network that will contain your application database.
```

> In the IPv4 address you should have a default value of 10.0.0.0/16 (otherwise add that network space to the box) and then add two subnets, one named public, with the subnet address range of 10.0.1.0/24 and one named private with the subnet range of 10.0.2.0/24.

Reviewer: ```suggestion
In the IPv4 address, you should have a default value of 10.0.0.0/16 (otherwise add that network space to the box). Next, add two subnets:

- one subnet named public, with the subnet address range of 10.0.1.0/24, and
- one subnet named private with the subnet range of 10.0.2.0/24.
```

![Azure network configuration](images/AZNetworks.png)

> Hit **Next** button till you get to the **Review and create** page where you can click on the create button at the bottom to start the creation of your network.

Reviewer: ```suggestion
Click **Next** until you get to the **Review and create** page, then click on the **Create** button at the bottom to create your network.
```

### Step 3: Create the database

> Now head back to the Resource Group you created or the Azure Management console homepage, and create a new resource. This time search for **Azure Database for PostgreSQL**. You can find this one more easily by selecting **Databases** in the Marketplace menu.

Reviewer: This paragraph could be clarified: where should I go to create the database?
```suggestion
From the Azure Management console homepage, create a new resource. This time search for **Azure Database for PostgreSQL**. You can also find this database in the Marketplace menu under **Databases**.
```

![Azure PostgreSQL](images/AZDatabase.png)

In the next page you will have to select the way you will use the service. Choose **Single server** for now, which is a good starting point, but take into account that this will create just one instance of PostgreSQL that will be your single point of failure.

> On the next screen, select or enter the following:

Reviewer: If one item in this list ends with a period, they all should end with a period.

> On the next screen, select or enter the following:
> - Resource group: the one you are including all your components

Reviewer: ```suggestion
On the next screen, select or enter the following:

- Resource group: the one you are including all your components
```

- Server name: a unique name for your database
- Data Source: can be left as `None`
- Location: the same one you used for your Resource Group and VNET
- Version: use the latest one you can

> - Compute + Storage: you can re-dimension your database but it's mandatory to select a `General Purpose` tier as it's the only that provides a Private Link at the time of the writing of this guide.

Reviewer: ```suggestion
- Compute + Storage: you can re-dimension your database,  but you must select the `General Purpose` tier, as it's the only tier that provides a Private Link.
```

- choose an admin username and password of your choice

> Click **Next** till you get to the final page where you can click on **Create** button so your database starts the creation process. Now, you need to wait till the creation of the database is over so you can include it in your VNET and create the internal connection to it. Once the creation is over, click on the **Go to resource** button that appears on the page.

Reviewer: ```suggestion
Click **Next** until you get to the final page, then click **Create**. It'll take some time for Azure to create the database. Once the creation is complete, click on the **Go to resource** button.
```

### Step 4: Create the private endpoint connection

> On the properties page for the Azure for PostgreSQL database, you will be able to manage all parameters about your database. Here we will create the **Private Endpoint** to the database, so all traffic from the web application is routed internally through Azure's network.
> 

Reviewer: ```suggestion
On the properties page for the Azure for PostgreSQL database, you can manage parameters for your database. You'll create the **Private Endpoint** to the database, so all traffic from the web application is routed internally through Azure's network.

```

Now go to the next step where you will select:
- Options: single container
- Image source: DockerHub
- Access Type: Public
- Image and tag: metabase/metabase:latest (or choose any other docker image tag of your preference, like our Enterprise Edition that has super-powers). If you want to see which is the latest Metabase version you can check in our [Community Edition Dockerhub repository](https://hub.docker.com/r/metabase/metabase/tags?page=1&ordering=last_updated) and also our [Enterprise Edition Dockerhub Repository](https://hub.docker.com/r/metabase/metabase-enterprise/tags?page=1&ordering=last_updated)
- Startup command: empty

Click next till you get to the last section and click on **Create** to initialize the creation of your application and wait till it's over.

Now go to te application configuration page and click on Settings --> Networking on the left side of the page, where a page will appear and you will have to click on **Click here to configure** under **VNET integration**.

![Azure VNET integration](images/AZVNETintegration.png)

> Now click on the huge plus sign next to **Add VNET** and select the VNET that you created on the first step and the public subnet to set the server on the public subnet of the VNET, and click OK.

Reviewer 2: check out the new commit, let me know if you want to do a 1:1 to improve it (this step is specially tricky)

![Azure VNET public subnet](images/AZVNETPublicSubnet.png)

> Now go back to te application configuration page and click on Settings --> Configuration on the left side of the page where you will see a few Application Settings already configure.

Reviewer: ```suggestion
Return to the application configuration page and click on **Settings** -> **Configuration** on the left side of the page. You should see a few Application Settings already configured.
```

Here you will need to add the [Enviroment Variables]() for connecting Metabase to its [PostgreSQL Application Database](), but make sure that you use the **MB_DB_CONNECTION_URI** as PostgreSQL in Azure is configured by default with SSL so currently the only way of passing those parameters is by using the full URI.

Also, take into account that the username in Azure PostgreSQL is `user@name_of_your_database_engine` so in this case the entire connection uri would be as follows: `postgresql://databasePrivateIPAddress:port/postgres?user=user@name_of_your_database_engine&password=configuredpassword&ssl=true&sslmode=required`.

> Lets put an example, if you got this information from the previous steps:
> 1) database private IP address: 10.0.2.4
> 2) database port: 5432 (in the case of Postgres, MySQL/MariaDB default port is 3306)
> 3) name of the database server: metabase-app-database
> 4) username of the database: metabase
> 5) password: Password1!

Reviewer: ```suggestion
For example, if your values are:

1) **database private IP address**: 10.0.2.4
2) **database port**: 5432 (in the case of Postgres, MySQL/MariaDB default port is 3306)
3) **name of the database server**: metabase-app-database
4) **username of the database**: metabase
5) **password**: Password1!
```

> then my connection string will be: `postgresql://10.0.2.4:5432/postgres?user=metabase@metabase-app-database&password=Password1!&ssl=true&sslmode=require`

Reviewer: ```suggestion
Then your connection string would be: 
```
postgresql://10.0.2.4:5432/postgres?user=metabase@metabase-app-database&password=Password1!&ssl=true&sslmode=require
```
```

Reviewer: The connection string should be on it's own line.

> Click on save and the instance will restart. After this step you will be able to access your Metabase in the URL that you have in the "Overview" tab in the web app under the URL section (will be the name you chose in the first step of the config of the web app + .azurewebsites.net as Azure handles all the deployment and domain)

Reviewer: ```suggestion
Click **Save** and the instance will restart. 

Once it finishes, you should be able to visit your Metabase at the URL shown in the the "Overview" tab in the web app (under the URL section). 
```

## Additional configurations
### How to enable Health checks

> Enabling health checking in Metabase is a good practice. You can tell the "host" when Metabase is ready to server your customers by just doing a simple configuration: go to your web app -> Monitoring -> Health Check -> Enable health check and include in the path `/api/health`.

Reviewer: ```suggestion
Enabling health checking in Metabase is a good practice.  Go to your **web app** -> **Monitoring** -> **Health Check** -> **Enable health check**, and include in the path `/api/health`.
```

> ### How to update

Reviewer: This section should be titled "How to upgrade". And the section seems a bit thin. Is this all there is to it? Can we point them to our upgrade guide?

Simply go to your web app and in Settings -> Container Settings (Classic) and there change the image version of Metabase.
In the same way you do an upgrade, you can change the Metabase version to Enterprise by switching to v1 instead of v0 in the prefix of the docker image.

**Important**: always ensure you have a backup of your Database before doing a version upgrade, *specially*, when upgrading between major versions.

### How to see the logs

> If you want to see the logs of your deployment, you need to go to the web app that you created and you will be able to see the streaming logs in Monitoring -> Log stream.

Reviewer: ```suggestion
Visit your web app in Azure, and navigate to **Monitoring -> Log stream**.
```
Do we want to point them to the troubleshooting logs in the Metabase Admin panel?

### Performance tuning

- Disable FTP State
- Change HTTP version to 2.0
- Enable WebSockets
- ARR Affinity enabled in case you scale out

### Custom domains

> You can go to Settings -> custom domains in your web app use a custom domain that you have in Azure, with your own certificates

Reviewer: ```suggestion
In your Azure web app, visit **Settings** -> **Custom domains** to use a custom domain with your own certificates.
```

### Scale-up vs Scale-out

> Azure provides an easy way in which you can increase the server's horsepower by doing a Scale-up. Simply choose a pricing tier with more RAM and CPU count and restart the instance. This approach is easier than scaling out and you should choose this option first before scaling out.
> 
> In case you know what you are doing and want to have a highly-available Metabase deployment, you can tell your web app to scale-out (more servers serving the application in parallel). Simply go to Scale-out and use the slider to ask for more instances or use a custom autoscaling policy (like more servers during your work hours and less during night, or anything you come up with)

Reviewer: ```suggestion
Azure provides an easy way to increase the server's capacity by doing a Scale-up. Simply choose a pricing tier with more RAM and CPU count and restart the instance. This approach is easier than scaling out and you should choose this option first before scaling out.

For a highly-available Metabase deployment, you can tell your web app to scale-out (more servers serving the application in parallel). Simply go to Scale-out and use the slider to ask for more instances, or use a custom autoscaling policy (like using more servers during your work hours and fewer during off hours).
```

### CORS

> In case you are using embedding, you might have to enable CORS in Settings -> CORS in the web application. You will need to fill the origin domains from where you are embedding so Azure knows that they need to enable cross requests from there.

Reviewer: ```suggestion
In case you're embedding Metabase, you might need to enable CORS in **Settings** -> **CORS** in the web application. You'll need to fill the origin domains for the app with the embedded Metabase, so Azure knows to enable cross requests from those domains.
```

### Database name

Azure does not let users create a database upon service creation, that's the reason why we used `postgres` as the database to install Metabase. Althought this shouldn't be a problem, a good practice would be to install the database in a separate database named `metabase`. If you are not in a hurry to try the product, you should create a database named `metabase` as soon as you create the database and then use the appropiate connection string when deploying the docker container.

> In the example above the connection string would be `postgresql://10.0.2.4:5432/metabase?user=metabase@metabase-app-database&password=Password1!&ssl=true&sslmode=require`

Reviewer: I'm not sure I understand this section. Why is the naming a good practice? And is the section necessary? What do we lose if we cut it?
